### ISSUE TITLE
randomily crashing

### ISSUE THREAD

**Issue by XtremeForce** [WilliamSchack/Spotify-Downloader#10]
in the middle of downloading it crashes randomly in average five minutes

**Comment by WilliamSchack** [comment]
I am sorry that that is happening and hopefully I can help you fix it

The only thing I can think of is that the normalisation process and having multiple threads can be a strain to less powerful cpus. Maybe try and turning off audio normalisation and change the downloader threads to 1 and see if that fixes the problem.

If not, do you know what stage the download was on when it crashed? Just because this is an issue I have not seen on my machine and it will help me narrow down debugging if there are any issues I need to fix.

Hopefully lowering the settings works but if your cpu is capable of running the multiple threads and ffmpeg, we can gladly look into it further.

**Comment by WilliamSchack** [comment]
Other than that there may also be a stage in the downloading process that is going wrong, so if lowering settings doesnt work, would you please be able to set the downloading threads to one, turn on audio normalisation, and let me know the song you used, and what stage the downloader crashed?

**Comment by XtremeForce** [comment]
> Owner
i have a stong cpu that can handle heavy process
i have even put the threads to 6 then crashed again then 12 and it crashes faster by putting more threads even i turned off normalizing the song once and crashed

**Comment by XtremeForce** [comment]
im trying now with one thread and normalizing off

**Comment by XtremeForce** [comment]
![image](https://github.com/user-attachments/assets/7fca154b-c822-4bd3-9455-2e030b23a42e)
it crashed here while in one thread and no normalize volume
i see the download go further when the threads are down but it still crashes after time

**Comment by XtremeForce** [comment]
![image](https://github.com/user-attachments/assets/789bde83-5457-4b11-9805-0095ab426f80)
it crashed here exactly

**Comment by WilliamSchack** [comment]
I see, thats most likely something different with the data of that song, sometimes songs on spotify can have data that the program doesn't like for whatever reason, Ill try debugging it with that song and see if I can fix it, please note though that I will be busy for the fair chunk of today and I am not at my computer so it may be a day or so before its fixed, sorry about that.

Also, was there any other songs having this issue? Or does it just crash on this one, or like you said before it it just random songs?

**Comment by WilliamSchack** [comment]
Wait you said something about the threads, what do you mean by that?

**Comment by WilliamSchack** [comment]
Oh right before you said that it crashes quickly with more threads. I don't think its an issue with the threads here because if it failed at the metadata, it is most likely having problem with taglib setting the tags of a song, so the larger amount of threads may just be getting to that specific song, or other problematic songs quicker as it is downloading songs at a quicker rate

**Comment by XtremeForce** [comment]
> I see, thats most likely something different with the data of that song, sometimes songs on spotify can have data that the program doesn't like for whatever reason, Ill try debugging it with that song and see if I can fix it, please note though that I will be busy for the fair chunk of today and I am not at my computer so it may be a day or so before its fixed, sorry about that.
> 
> Also, was there any other songs having this issue? Or does it just crash on this one, or like you said before it it just random songs?

nvm   take your time
it doesnt crash at a specific song as the playlist download random song sometimes
about the threads as i raise the number of threads the program crashes faster at one thread it crashed at the 150 song at 3 threads first it crashed at the 35 and at 6 threads at 60 and at 12 at 85 and at 16 it crashed right now
i dont remeber the specific number and threads but i am sure that it crashes with more threads

**Comment by WilliamSchack** [comment]
That is strange, I understand if you don't want to but would it be possible if I use your playlist for debugging this? Just so I am using the same songs that you used to narrow down the issue further. This is me thinking ahead though, so it could be a problem could switching songs as the metadata status will stay until the next song is set up and ready to start downloading.

**Comment by XtremeForce** [comment]
> Oh right before you said that it crashes quickly with more threads. I don't think its an issue with the threads here because if it failed at the metadata, it is most likely having problem with taglib setting the tags of a song, so the larger amount of threads may just be getting to that specific song, or other problematic songs quicker as it is downloading songs at a quicker rate

i see that the numbers in the album cover and the ones in the bar where the pause at
it shows at 3 threads no/83 and at 6 threads no/41

**Comment by XtremeForce** [comment]
> That is strange, I understand if you don't want to but would it be possible if I use your playlist for debugging this? Just so I am using the same songs that you used to narrow down the issue further. This is me thinking ahead though, so it could be a problem could switching songs as the metadata status will stay until the next song is set up and ready to start downloading.

sure u can take my playlist here it is https://open.spotify.com/playlist/2Q94TnsTIn3ixFUbpSPKhh?si=a730a2a9729c4f51

**Comment by WilliamSchack** [comment]
> i see that the numbers in the album cover and the ones in the bar where the pause at it shows at 3 threads no/83 and at 6 threads no/41

Yeah that sounds like it its a song that is having problems because the song count is halving as you double the threads, which should run at double the speed. I will definitely test the threads though, I never tested really high values like above 10 before since when I implemented it I had a pretty poor cpu, but it should be good to handle it now so I will see if there is any issues with it.

**Comment by XtremeForce** [comment]
> > i see that the numbers in the album cover and the ones in the bar where the pause at it shows at 3 threads no/83 and at 6 threads no/41
> 
> Yeah that sounds like it its a song that is having problems because the song count is halving as you double the threads, which should run at double the speed. I will definitely test the threads though, I never tested really high values like above 10 before since when I implemented it I had a pretty poor cpu, but it should be good to handle it now so I will see if there is any issues with it.

12 threads was frying my pc really but it was working

**Comment by WilliamSchack** [comment]
> 12 threads was frying my pc really but it was working

Probably keep the thread count down from now I really dont want to kill your pc

**Comment by XtremeForce** [comment]
> > 12 threads was frying my pc really but it was working
> 
> Probably keep the thread count down from now I really dont want to kill your pc

my pc can handle this but its like to run minecraft in rtx mod

**Comment by WilliamSchack** [comment]
> my pc can handle this but its like to run minecraft in rtx mod

Yeah it is a bit taxing just because the downloading of the song takes a good amount of cpu usage, but when normalising, ffmpeg eats a lot of the cpu and having it ongoing on 12 different threads will really hurt the performance of your pc

**Comment by XtremeForce** [comment]
![image](https://github.com/user-attachments/assets/d4287474-229f-440a-a2b8-4094400eec7a)
at 12 threads it crashed here it look like it crashes at the asiiginig metadata phase

**Comment by XtremeForce** [comment]
> > my pc can handle this but its like to run minecraft in rtx mod
> 
> Yeah it is a bit taxing just because the downloading of the song takes a good amount of cpu usage, but when normalising, ffmpeg eats a lot of the cpu and having it ongoing on 12 different threads will really hurt the performance of your pc

as it takes less time i would leave the pc till it fully downloads

**Comment by XtremeForce** [comment]
as i tested any thread above 12 crashes the program

**Comment by WilliamSchack** [comment]
> as i tested any thread above 12 crashes the program

yeah that sounds about right, its just using a lot of cpu and it most likely cant handle it. Please turn down the thread count

**Comment by XtremeForce** [comment]
> > as i tested any thread above 12 crashes the program
> 
> yeah that sounds about right, its just using a lot of cpu and it most likely cant handle it. Please turn down the thread count

6 threads seems good

**Comment by WilliamSchack** [comment]
> 6 threads seems good

As long as its the only thing running on your pc, I would personally recommend 3 just to minimise the usage of it and to allow for using other programs while its running in the background. But if its running fine for you then 6 threads should be good

**Comment by XtremeForce** [comment]
i think i found the bug
there are some songs in my playlist that isnt in spotify as i watched the program crashes when downloading it as it doesnt found any thing

**Comment by XtremeForce** [comment]
so if you can add skip to the songs that arent found it would solve the problem

**Comment by WilliamSchack** [comment]
> so if you can add skip to the songs that arent found it would solve the problem

Yeah I was going to implement that later just to avoid this issue in the future. I will still debug this issue though to hopefully fix a few more songs from errors

**Comment by WilliamSchack** [comment]
Sorry about the wait as well you caught me when I woke up on my busiest day of the week so I will be able to work on it in about 8 hours from now

**Comment by XtremeForce** [comment]
> > so if you can add skip to the songs that arent found it would solve the problem
> 
> 
> 
> Yeah I was going to implement that later just to avoid this issue in the future. I will still debug this issue though to hopefully fix a few more songs from this error

after deleting all the songs the program continued normaly although i runned it at 12 thread
after that there were songs not found by the program appeared to me

**Comment by XtremeForce** [comment]
> Sorry about the wait as well you caught me when I woke up on my busiest day of the week so I will be able to work on it in about 8 hours from now

never mind it is a very good program that helped me so much 
i searched for plenty of spotify downloaders and all of them was bad 
you are doing a great job ❤️

**Comment by WilliamSchack** [comment]
> after deleting all the songs the program continued normaly although i runned it at 12 thread after that there were songs not found by the program appeared to me

That is really strange I will keep that in mind

**Comment by WilliamSchack** [comment]
Fixed in commit https://github.com/WilliamSchack/Spotify-Downloader/commit/f01aadeed7bfa9d1e085cd8261e927e3dbabffa9 and will be released very soon :)

So this was a very strange problem but it turned out that when updating the progress bar, if the song produced a file name that included the "%" symbol, and yt-dlp notified me that it was modifying the file, it would crash the program due to it not being the output I read for the progress bar.

Also when debugging with high thread counts I also found out that near the end of downloading, a thread can request a UI update but be deleted before it is processed, crashing the program, which is now fixed, so thanks for that as well!

This should have fixed your issue now and hopefully there are minimal bugs in future use!
